### PR TITLE
Display optimization progress for find_MAP

### PR DESCRIPTION
**Pull request by bwengals**
If `disp=True` in call to `find_MAP`, then optimization progress is printed out.
- If ipython notebook is being used, an updating table showing parameter progress is displayed.  Colors are taken from PyMC3 logo.
- If ipython notebook is not being used, more basic info (number of iterations, log posterior value, ||grad||, runtime) is displayed every 2 seconds.

Also in this PR is a small change to appearances of tqdm.  If the user is developing in a notebook, the better looking tqdm notebook iteration bar is displayed.

Let me know if there are any changes, cosmetic or otherwise, you would like or would prefer.

terminal version

notebook version

### CODE REVIEW

File: pymc3/tests/test_starting.py [hunk 1]
```diff
@@ -80,3 +80,22 @@ def test_find_MAP():
...
+        mu = Normal('mu', 0, 1, testval=100)
+        map_est1 = starting.find_MAP(fmin=starting.optimize.fmin_bfgs,   disp=False)
+        map_est2 = starting.find_MAP(fmin=starting.optimize.fmin_powell, disp=False)
+
+    # providing a callback should override disp (since it is a callback)
+    #   should not error
+    with Model():
+        mu = Normal('mu', 0, 1, testval=100)
+        # test with verbosity turned off (it is on by default)
+        map_est1 = starting.find_MAP(fmin=starting.optimize.fmin_bfgs,   disp=True, callback = lambda x: 1.0)
```
Comment (twiecki) [inline pymc3/tests/test_starting.py]: no space for kwargs

File: pymc3/tests/test_starting.py [hunk 2]
```diff
@@ -80,3 +80,22 @@ def test_find_MAP():
...
     close_to(map_est2['mu'], 0, tol)
     close_to(map_est2['sigma'], 1, tol)
+
+
+def test_find_MAP_noverbosity():
+    # test with verbosity turned off (it is on by default)
+    #   should not error
+    with Model():
+        mu = Normal('mu', 0, 1, testval=100)
+        map_est1 = starting.find_MAP(fmin=starting.optimize.fmin_bfgs,   disp=False)
```
Comment (twiecki) [inline pymc3/tests/test_starting.py]: no extra spacing

File: pymc3/tuning/starting.py [hunk 3]
```diff
@@ -171,3 +174,154 @@ def allinmodel(vars, model):
...
+            html += r"""
+              <tr>
+                <td class="tg-bgft">{}</td>
+                <td class="tg-vkoh">{}</td>
+                <td class="tg-vkoh">{}</td>
+              </tr>
+            """.format(parameter, values["size"], values["valstr"])
+        html += "</table>"
+        self.param_table.value = html
+
```
Comment (twiecki) [inline pymc3/tuning/starting.py]: empty-lines

**Comment by twiecki** [comment]
Do you have a screenshot of what the Monitor in find_MAP() looks like?

File: pymc3/tests/test_starting.py (hunk 1 above)
Comment (bwengals) [reply pymc3/tests/test_starting.py]: what do you mean?

Comment (twiecki) [reply pymc3/tests/test_starting.py]: should be `callback=lambda x: 1.0`

Comment (twiecki) [reply pymc3/tests/test_starting.py]: and `fmin=starting.optimize.fmin_bfgs, disp=True`

**Comment by bwengals** [comment]
@twiecki added screenshots to the original comment.  One thing you'll notice is that values of untransformed variables are printed.  Is there a clean way to transform them back?  I had a hard time finding that poking around.

**Comment by twiecki** [comment]
I'm not sure about Monitor. find_MAP() is not really a strong focus of this package. We could use it in other parts too (MCMC) but not sure how useful that really is.

Then I think there is the problem of creating a lot of html output in a cell. If you look at the raw NB after an optimization run, does it have a huge amount of html code in it?

**Comment by bwengals** [comment]
It looks like the monitor does not persist across saves.  Since it uses the interactive widgets, one needs to [do this](http://ipywidgets.readthedocs.io/en/latest/embedding.html) to save it.

And additionally, I imagine a version for MCMC chains could get messy pretty fast.  I guess my arguments in favor would be that finding the MAP point is a good first step in (not all!) but many situations, like GPs.  It is helpful (for me at least) to understand subtle problems in model specification by seeing the parameters veer off somewhere, without waiting around till it terminates.  But very true, it's not an optimization package!

Could dial back the css to not seem so "branded", since its not the focus.

**Comment by fonnesbeck** [comment]
I'm not necessarily opposed, as long as it stays relatively lightweight. I do think more sophisticated monitoring would be nice for NUTS and ADVI. One could make it optional, with a `monitor=True` flag, or similar. 

I've thought about using Bokeh for some graphical monitoring, similar to what dask does for its dashboard. But you wouldnt necessarily want it all the time.

**Comment by bwengals** [comment]
Could do something similar for the samplers and ADVI.  For something lightweight, maybe it would look similar but show marginal mean and standard deviation in the "Current Value" column? 

Yeah, default should be `monitor=False`.  Like [this?](https://raw.githubusercontent.com/dask/dask-org/master/images/daskboard.gif)

File: pymc3/sampling.py [hunk 4]
```diff
@@ -187,7 +187,11 @@ def _sample(draws, step=None, start=None, trace=None, chain=0, tune=None,
     sampling = _iter_sample(draws, step, start, trace, chain,
                             tune, model, random_seed)
     if progressbar:
-        sampling = tqdm(sampling, total=draws)
+        try:
```
Comment (twiecki) [inline pymc3/sampling.py:190]: We should refactor this to our own `tqdm` function which does this.

File: pymc3/tests/test_starting.py [hunk 5]
```diff
@@ -80,3 +80,22 @@ def test_find_MAP():
...
+        mu = Normal('mu', 0, 1, testval=100)
+        map_est1 = starting.find_MAP(fmin=starting.optimize.fmin_bfgs, disp=False)
+        map_est2 = starting.find_MAP(fmin=starting.optimize.fmin_powell, disp=False)
+
+    # providing a callback should override disp (since monitor works via callback to sp.optimize)
+    #   should not error
+    with Model():
+        mu = Normal('mu', 0, 1, testval=100)
+        # test with verbosity turned off (it is on by default)
+        map_est1 = starting.find_MAP(fmin=starting.optimize.fmin_bfgs, disp=True, callback = lambda x: 1.0)
```
Comment (twiecki) [inline pymc3/tests/test_starting.py:98]: no white-space around callback kwarg =.

File: pymc3/tuning/starting.py [hunk 6]
```diff
@@ -171,3 +174,134 @@ def allinmodel(vars, model):
     notin = [v for v in vars if v not in model.vars]
     if notin:
         raise ValueError("Some variables not in the model: " + str(notin))
+
+
+
+class Monitor(object):
```
Comment (twiecki) [inline pymc3/tuning/starting.py:180]: Does there exist nothing in  the jupyter (or bokeh) universe that does this already?

Comment (bwengals) [reply pymc3/tuning/starting.py:180]: Gosh, not that I've seen.  Couldn't find anything similar after a quick look around, but I'll check further.  I took the idea of monitoring optimization progress from GPy, theirs was made from scratch also.

File: pymc3/sampling.py (hunk 4 above)
Comment (bwengals) [reply pymc3/sampling.py:190]: That sounds good.  Where would you recommend it goes?

**Comment by aseyboldt** [comment]
I think a couple of calls to `tqdm.close()` are missing. Otherwise it sometimes prints weird things into the next output in the notebook if the iterator is not exhausted (because of an exception or a keyboard interrupt). Something like this should do it:
```python
vals = tqdm.tqdm(some_iter)
try:
    for val in vals:
        pass
finally:
    vals.close()
```
Or alternatively using `contextmanager.closing`.
Most of them were missing before this patch...

File: pymc3/sampling.py [hunk 7]
```diff
@@ -401,7 +405,11 @@ def sample_ppc(trace, samples=None, model=None, vars=None, size=None, random_see
     seed(random_seed)
 
     if progressbar:
-        indices = tqdm(randint(0, len(trace), samples), total=samples)
+        try:
```
Comment (aseyboldt) [inline pymc3/sampling.py:408]: close missing

File: pymc3/variational/advi.py [hunk 8]
```diff
@@ -148,10 +147,15 @@ def advi(vars=None, start=None, model=None, n=5000, accurate_elbo=False,
     elbos = np.empty(n)
     divergence_flag = False
     try:
-        progress = trange(n)
         uw_i, elbo_current = f()
         if np.isnan(elbo_current):
             raise FloatingPointError('NaN occurred in ADVI optimization.')
+        # set up progress bar
+        try:
```
Comment (aseyboldt) [inline pymc3/variational/advi.py:154]: missing close

File: pymc3/variational/advi.py [hunk 9]
```diff
@@ -400,7 +404,14 @@ def rvs(x):
     trace = pm.sampling.NDArray(model=model, vars=vars_sampled)
     trace.setup(draws=draws, chain=0)
 
-    range_ = trange(draws) if progressbar else range(draws)
+    if progressbar:
```
Comment (aseyboldt) [inline pymc3/variational/advi.py:407]: missing close

File: pymc3/variational/advi_minibatch.py [hunk 10]
```diff
@@ -518,9 +518,14 @@ def is_shared(t):
     updates = OrderedDict(optimizer(loss=-1 * elbo, param=params))
     f = theano.function(tensors, elbo, updates=updates, mode=mode)
 
+    # set up progress bar
+    try:
```
Comment (aseyboldt) [inline pymc3/variational/advi_minibatch.py:522]: missing close

**Comment by bwengals** [comment]
Thanks @aseyboldt, nice catch.  Will include this.

**Comment by bwengals** [comment]
Want to give a heads up that, in addition to the suggestions, I'm working on having it print **un**transformed variables, ie. the value of  `sigma` instead of `sigma_log_` .

**Pull request closed without merging**